File: src/main/python/AoC2024_25.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 25
#

import sys

from aoc import my_aocd
from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Heights = tuple[int, ...]
Input = tuple[set[Heights], set[Heights]]
Output1 = int
Output2 = str


TEST = """\
#####
.####
.####
.####
.#.#.
.#...
.....

#####
##.##
.#.##
...##
...#.
...#.
.....

.....
#....
#....
#...#
#.#.#
#.###
#####

.....
.....
#.#..
###..
###.#
###.#
#####

.....
.....
.....
#....
#.#..
#.#.#
#####
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        keys, locks = set(), set()
        blocks = my_aocd.to_blocks(input_data)
        for block in blocks:
            rows = ["".join(x) for x in zip(*block, strict=True)]
            nums: Heights = tuple(
                sum(1 for ch in row if ch == "#") - 1 for row in rows
            )
            if rows[0][0] == "#":
                locks.add(nums)
            else:
                keys.add(nums)
        return keys, locks

    def part_1(self, schematic: Input) -> Output1:
        keys, locks = schematic
        return sum(
            all(a + b <= 5 for a, b in zip(lock, key, strict=True))
            for lock in locks
            for key in keys
        )

    def part_2(self, _schematic: Input) -> Output2:
        return "🎄"

    @aoc_samples((("part_1", TEST, 3),))
    def samples(self) -> None:
        pass


solution = Solution(2024, 25)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_25.java
import static com.github.pareronia.aoc.itertools.IterTools.enumerate;
import static com.github.pareronia.aoc.itertools.IterTools.product;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.List;

public class AoC2024_25 extends SolutionBase<AoC2024_25.Input, Long, String> {

    private AoC2024_25(final boolean debug) {
        super(debug);
    }

    public static final AoC2024_25 create() {
        return new AoC2024_25(false);
    }

    public static final AoC2024_25 createDebug() {
        return new AoC2024_25(true);
    }

    @Override
    protected Input parseInput(final List<String> inputs) {
        return Input.fromInputs(inputs);
    }

    @Override
    public Long solvePart1(final Input input) {
        log(input);
        return product(input.keys, input.locks).stream()
                .filter(pp -> (pp.first() & pp.second()) == 0)
                .count();
    }

    @Override
    public String solvePart2(final Input input) {
        return "🎄";
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "3"),
    })
    public void samples() {}

    public static void main(final String[] args) throws Exception {
        AoC2024_25.create().run();
    }

    private static final String TEST =
            """
            #####
            .####
            .####
            .####
            .#.#.
            .#...
            .....

            #####
            ##.##
            .#.##
            ...##
            ...#.
            ...#.
            .....

            .....
            #....
            #....
            #...#
            #.#.#
            #.###
            #####

            .....
            .....
            #.#..
            ###..
            ###.#
            ###.#
            #####

            .....
            .....
            .....
            #....
            #.#..
            #.#.#
            #####
            """;

    record Input(List<Long> keys, List<Long> locks) {

        public static Input fromInputs(final List<String> inputs) {
            final List<Long> ks = new ArrayList<>();
            final List<Long> ls = new ArrayList<>();
            for (final List<String> block : StringOps.toBlocks(inputs)) {
                final int h = block.size();
                final long n =
                        enumerate(block.stream().skip(1)).stream()
                                .flatMapToLong(
                                        e -> {
                                            final int r = e.index();
                                            final String line = e.value();
                                            return enumerate(Utils.asCharacterStream(line)).stream()
                                                    .filter(ee -> ee.value() == '#')
                                                    .mapToLong(ee -> 1L << ((h * ee.index()) + r));
                                        })
                                .sum();
                if (block.get(0).charAt(0) == '#') {
                    ls.add(n);
                } else {
                    ks.add(n);
                }
            }
            return new Input(ks, ls);
        }
    }
}
